### ISSUE TITLE
Suggestion: Generators

### ISSUE THREAD

**Issue by wavebeem** [minamorl/trishula#19]
I don't know a lot about Python generators, but I wonder if you could use them for parsing?

```py
@T.Generator
def json_array():
    yield T.Value("[")
    yield json_ws
    items = yield sep_by(json_value(), json_sep())
    yield json_ws
    yield T.Value("]")
    return T.Succeed(items)
```

**Comment by minamorl** [comment]
I had same idea but as far as I searched, Python do not allow combination of yield and return unfortunately.

**Comment by wavebeem** [comment]
It could be possible with yield still I think, but I don't know how to do it right now

**Comment by minamorl** [comment]
`items = yield sep_by(json_value(), json_sep())` is the most difficult point, because `someGenerator.send(value)` sends the value at first, then returns yielded value...

**Comment by minamorl** [comment]
Maybe this can be done with calling a generator twice. Firstly you collect grammars, then send the results back to the generator. But this seems tricky solution.

**Comment by minamorl** [comment]
I consider something like this:
```python
import trishula as T
import operator

def generator_wrapper(generator):
    # to receive first sended value
    yield
    yield from generator

def parser():
    yield T.Value("1")
    value = yield T.Value("2")
    yield T.Value("3")
    # return value as a result
    yield value


def collect_grammars(generator):
    next(generator)
    grammars = []
    for x in generator:
        grammars.append(x)
    grammars = grammars[:-1]
    return grammars

def send_back(generator, results):
    next(generator)
    x = None
    for r in results:
        x = generator.send(r)
    # retrive final value
    return x

def parse(s):
    a = generator_wrapper(parser())
    b = generator_wrapper(parser())
    grammars = collect_grammars(a)
    results = [None]
    current_index = 0
    for p in grammars:
        result = p.parse(s, current_index)
        current_index = result.index
        if not result.isSuccess():
            return result
        results.append(result.value)
    return send_back(b, results)


print(parse("123")) #==> 2
```

**Comment by minamorl** [comment]
Add `define_parser` function in trishula.decorator. https://github.com/minamorl/trishula/pull/25
Released as 0.0.8.

**Comment by wavebeem** [comment]
Nice
